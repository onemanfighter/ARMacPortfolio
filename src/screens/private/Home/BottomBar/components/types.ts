export enum ProgramType {
  FINDER = 'finder',
  CHROME = 'chrome',
  VSCODE = 'vscode',
  TERMINAL = 'terminal',
  SPOTIFY = 'spotify',
  GITHUB = 'github',
  SETTINGS = 'settings',
}

export type IconButtonProps = {
  type: ProgramType;
};
