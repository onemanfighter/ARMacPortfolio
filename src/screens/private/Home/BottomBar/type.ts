import { ProgramType } from './components';

export type BottomBarProps = {};
export type ProgramItem = { type: ProgramType; name: string; id: string };
