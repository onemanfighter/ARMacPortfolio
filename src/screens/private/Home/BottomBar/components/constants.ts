import { ProgramType } from './types';
import Finder from '../../../../../assets/logo/finder.png';
import Chrome from '../../../../../assets/logo/chrome.png';
import VsCode from '../../../../../assets/logo/vscode.png';
import Terminal from '../../../../../assets/logo/terminal.png';
import Spotify from '../../../../../assets/logo/spotify.png';
import Github from '../../../../../assets/logo/github.png';
import Settings from '../../../../../assets/logo/settings.png';
import Bin from '../../../../../assets/logo/bin.png';
import Notes from '../../../../../assets/logo/notes.png';

export const PROGRAMS: Record<ProgramType, { name: string; icon: string }> = {
  [ProgramType.FINDER]: {
    name: 'Finder',
    icon: Finder,
  },
  [ProgramType.NOTES]: {
    name: 'Notes',
    icon: Notes,
  },
  [ProgramType.CHROME]: {
    name: 'Chrome',
    icon: Chrome,
  },
  [ProgramType.VSCODE]: {
    name: 'VsCode',
    icon: VsCode,
  },
  [ProgramType.TERMINAL]: {
    name: 'Terminal',
    icon: Terminal,
  },
  [ProgramType.SPOTIFY]: {
    name: 'Spotify',
    icon: Spotify,
  },
  [ProgramType.GITHUB]: {
    name: 'Github',
    icon: Github,
  },
  [ProgramType.SETTINGS]: {
    name: 'Settings',
    icon: Settings,
  },
  [ProgramType.BIN]: {
    name: 'Bin',
    icon: Bin,
  },
};

export default PROGRAMS;
