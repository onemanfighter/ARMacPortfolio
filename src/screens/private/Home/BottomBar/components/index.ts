export { default as ProgramButton } from './ProgramButton';
export type { IconButtonProps } from './types';
export { ProgramType } from './types';
