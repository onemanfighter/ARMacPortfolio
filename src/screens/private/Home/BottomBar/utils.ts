import { ProgramType } from '@processStore';
import { ProgramItem } from './type';

export const reorder = (
  list: ProgramItem[],
  startIndex: number,
  endIndex: number,
) => {
  const result = Array.from(list);
  const [removed] = result.splice(startIndex, 1);
  result.splice(endIndex, 0, removed);

  return result;
};

export const getAllAppOtherThanBottomBarApps = (
  apps: (ProgramType | undefined)[],
) => {
  return apps.filter(
    (app) =>
      !(
        app === ProgramType.FINDER ||
        app === ProgramType.SETTINGS ||
        app === ProgramType.LAUNCHPAD ||
        app === ProgramType.NOTES ||
        app === ProgramType.CHROME ||
        app === ProgramType.VSCODE ||
        app === ProgramType.TERMINAL ||
        app === ProgramType.SPOTIFY ||
        app === ProgramType.GITHUB ||
        app === ProgramType.BIN
      ) && app !== undefined,
  ) as ProgramType[];
};
