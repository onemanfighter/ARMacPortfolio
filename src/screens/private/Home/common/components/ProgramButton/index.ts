export { default as ProgramButton } from './ProgramButton';
export { getIconSize } from './utils';
