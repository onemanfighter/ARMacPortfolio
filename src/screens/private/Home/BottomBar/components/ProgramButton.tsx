import { Box, IconButton, Img, Tooltip } from '@chakra-ui/react';
import { IconButtonProps } from './types';
import PROGRAMS from './constants';
import { displayDockSelector, settingsStore, useShallow } from '@settingsStore';
import { getIconSize } from './utils';
import { Draggable } from 'react-beautiful-dnd';

const ProgramButton = ({ isActive, type, index, id }: IconButtonProps) => {
  const { dockSize } = settingsStore(useShallow(displayDockSelector));
  const { name, icon } = PROGRAMS[type] ?? { name: 'Unknown', icon: 'unknown' };
  const size = getIconSize(dockSize);
  return (
    <Draggable draggableId={id} index={index} key={name + '-button'}>
      {(provided) => (
        <Box
          ref={provided.innerRef}
          {...provided.draggableProps}
          {...provided.dragHandleProps}
          display={'flex'}
          flexDir={'column'}
          alignItems={'center'}
        >
          <Tooltip
            borderRadius={6}
            label={name}
            hasArrow
            arrowSize={15}
            top={-4}
          >
            <Img
              boxSize={size}
              src={icon}
              height={size}
              width={size}
              alt={name}
              _active={{
                filter: 'brightness(0.6)',
              }}
              loading="lazy"
            />
          </Tooltip>
          {isActive ? (
            <Box boxSize={1} bgColor={'white'} borderRadius={'full'} />
          ) : null}
        </Box>
      )}
    </Draggable>
  );
};

export default ProgramButton;
