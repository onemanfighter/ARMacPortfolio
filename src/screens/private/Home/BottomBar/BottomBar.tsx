import { useContext, useEffect, useState } from 'react';
import { DragDropContext, Droppable, DropResult } from 'react-beautiful-dnd';
import { Box, Divider } from '@chakra-ui/react';
import { BottomBarProps, ProgramItem } from './type';
import { ProgramButtonDraggable } from './components';
import { getAllAppOtherThanBottomBarApps, reorder } from './utils';
import PROGRAMS from './components/constants';
import {
  darkModeColorSelector,
  settingsStore,
  useShallow,
} from '@settingsStore';
import {
  activeAppActionsSelector,
  activeAppRunningSelector,
  activeAppSelector,
  allActiveAppsSelector,
  processStore,
  ProgramType,
  WindowSize,
} from '@processStore';
import { LaunchpadContext } from '../../Mac';
import { ProgramButton } from '../common';
import { LaunchpadProgramsList } from '../Launchpad/constants';
import { BottomBarProgramType } from './components/types';

const InitialProgramOrderList: ProgramItem[] = Object.entries(PROGRAMS).map(
  (value) => ({
    type: value[0] as BottomBarProgramType,
    name: value[1].name,
    id: value[1].name,
  }),
);

const BottomBar = (_props: BottomBarProps) => {
  const { launchpad, setLaunchpad } = useContext(LaunchpadContext);
  const [state, setState] = useState<ProgramItem[]>(InitialProgramOrderList);
  const { bottomBarBgColor } = settingsStore(useShallow(darkModeColorSelector));
  const { addApp, setWindowSize } = processStore(
    useShallow(activeAppActionsSelector),
  );
  const { allActiveApps, getActiveApp, activeAppRunning } = processStore(
    useShallow((state) => ({
      getActiveApp: activeAppSelector(state),
      allActiveApps: allActiveAppsSelector(state),
      ...activeAppRunningSelector(state),
    })),
  );

  const runningMiddleApps = getAllAppOtherThanBottomBarApps(
    allActiveApps as ProgramType[],
  ).slice(-3);

  function onDragEnd(result: DropResult) {
    if (
      result.destination &&
      result.destination.index !== result.source.index
    ) {
      const newList = reorder(
        state,
        result.source.index,
        result.destination.index,
      );

      setState(newList);
    }
  }

  const binCurrentAppState =
    getActiveApp(BottomBarProgramType.FINDER as unknown as ProgramType) ??
    undefined;

  useEffect(() => {
    if (activeAppRunning !== ProgramType.LAUNCHPAD) {
      setLaunchpad(false);
      return;
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [activeAppRunning, setLaunchpad]);

  return (
    <Box
      aria-label="bottom-bar"
      w={'100%'}
      zIndex={900}
      display={'flex'}
      flexDirection={'column'}
      justifyContent={'center'}
      alignContent={'center'}
      alignItems={'center'}
    >
      <DragDropContext onDragEnd={onDragEnd}>
        <Droppable droppableId="droppable-list" direction="horizontal">
          {(provided) => (
            <>
              <Box
                ref={provided.innerRef}
                {...provided.droppableProps}
                display={'flex'}
                height={'auto'}
                p={1}
                my={1}
                mx={'auto'}
                position={'fixed'}
                alignItems={'center'}
                transition={'all 0.3s ease-in-out'}
                bg={bottomBarBgColor}
                borderRadius={'3xl'}
                border={'1px solid gray'}
                dropShadow={'md'}
                bottom={0}
                style={{ backdropFilter: 'blur(6px)' }}
                zIndex={10}
                gap={4}
              >
                {state.map((program, index) => {
                  const currentApp = getActiveApp(
                    program.type as unknown as ProgramType,
                  );
                  return (
                    <ProgramButtonDraggable
                      type={program.type}
                      key={program.name}
                      id={program.id}
                      isActive={
                        currentApp !== undefined ||
                        (program.type === BottomBarProgramType.LAUNCHPAD &&
                          launchpad === true)
                      }
                      index={index}
                      onClickHandler={(app: BottomBarProgramType) => {
                        if (app === BottomBarProgramType.LAUNCHPAD) {
                          setLaunchpad(!launchpad);
                          return; // Prevents opening app when launchpad is open
                        }
                        setLaunchpad(false); // Close launchpad when opening app
                        !currentApp
                          ? setTimeout(() => {
                              addApp(app as unknown as ProgramType);
                            }, 500)
                          : setWindowSize(
                              app as unknown as ProgramType,
                              WindowSize.DEFAULT,
                            );
                      }}
                    />
                  );
                })}
                {provided.placeholder}
                {runningMiddleApps.length !== 0 ? (
                  <>
                    <Divider orientation="vertical" height={' 16'} />
                    {runningMiddleApps.map((app) => {
                      const currentApp = getActiveApp(app);
                      return (
                        <ProgramButton
                          name={app.valueOf().toString()}
                          key={app}
                          variant="bottomBar"
                          icon={LaunchpadProgramsList[app].icon}
                          isActive={currentApp !== undefined}
                          onClickHandler={() => {
                            !currentApp
                              ? setTimeout(() => {
                                  addApp(app);
                                }, 500)
                              : setWindowSize(app, WindowSize.DEFAULT);
                          }}
                        />
                      );
                    })}
                  </>
                ) : null}
                <Divider orientation="vertical" height={'16'} />
                <ProgramButton
                  name="Bin"
                  variant="bottomBar"
                  icon={LaunchpadProgramsList['bin'].icon}
                  isActive={binCurrentAppState !== undefined}
                  onClickHandler={() => {
                    !binCurrentAppState
                      ? setTimeout(() => {
                          addApp(ProgramType.BIN);
                        }, 500)
                      : setWindowSize(ProgramType.BIN, WindowSize.DEFAULT);
                  }}
                />
              </Box>
            </>
          )}
        </Droppable>
      </DragDropContext>
    </Box>
  );
};

export default BottomBar;
