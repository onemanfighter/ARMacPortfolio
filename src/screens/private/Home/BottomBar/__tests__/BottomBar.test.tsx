import { fireEvent, render, screen } from '@testing-library/react';
import BottomBar from '../BottomBar';
import { renderHook } from '@testing-library/react-hooks';
import { processStore } from '@processStore';
import { ProgramType } from '../components';
import { LaunchpadContext } from '../../../Mac';

describe('BottomBar', () => {
  beforeEach(() => {
    const { result } = renderHook(() => processStore());
    result.current.ActiveApp.removeApp(ProgramType.FINDER);
    result.current.ActiveApp.removeApp(ProgramType.CHROME);
  });
  it('should render correctly to match snapshot', () => {
    const { container } = render(<BottomBar />);

    expect(container).toMatchSnapshot();
  });

  it('should render invoke onclick on launch pad click', async () => {
    const setLaunchpad = jest.fn();
    const { result } = renderHook(() => processStore());
    render(
      <LaunchpadContext.Provider value={{ launchpad: false, setLaunchpad }}>
        <BottomBar />
      </LaunchpadContext.Provider>,
    );

    fireEvent.click(screen.getByLabelText('program-button-launchPad'));
    await jest.advanceTimersByTimeAsync(1000);

    expect(setLaunchpad).toHaveBeenCalled();
  });

  it('should render correctly with props to match snapshot', async () => {
    const { result } = renderHook(() => processStore());
    render(<BottomBar />);

    fireEvent.click(screen.getByLabelText('program-button-finder'));
    await jest.advanceTimersByTimeAsync(1000);

    expect(result.current.ActiveApp.apps[ProgramType.FINDER]).toMatchSnapshot();
  });

  it('should render correctly finder app to match snapshot', async () => {
    const { result } = renderHook(() => processStore());
    render(<BottomBar />);

    fireEvent.click(screen.getByLabelText('program-button-finder'));
    await jest.advanceTimersByTimeAsync(1000);

    expect(result.current.ActiveApp.apps[ProgramType.FINDER]).toMatchSnapshot();
  });

  it('should render correctly chrome app to match snapshot', async () => {
    const { result } = renderHook(() => processStore());
    render(<BottomBar />);

    fireEvent.click(screen.getByLabelText('program-button-chrome'));
    await jest.advanceTimersByTimeAsync(1000);

    expect(result.current.ActiveApp.apps[ProgramType.CHROME]).toMatchSnapshot();
  });

  it('should render correctly spotify app to match snapshot', async () => {
    const { result } = renderHook(() => processStore());
    render(<BottomBar />);

    fireEvent.click(screen.getByLabelText('program-button-spotify'));
    await jest.advanceTimersByTimeAsync(1000);

    expect(
      result.current.ActiveApp.apps[ProgramType.SPOTIFY],
    ).toMatchSnapshot();
  });

  it('should render correctly terminal app to match snapshot', async () => {
    const { result } = renderHook(() => processStore());
    render(<BottomBar />);

    fireEvent.click(screen.getByLabelText('program-button-terminal'));
    await jest.advanceTimersByTimeAsync(1000);

    expect(
      result.current.ActiveApp.apps[ProgramType.TERMINAL],
    ).toMatchSnapshot();
  });

  it('should render correctly vscode app to match snapshot', async () => {
    const { result } = renderHook(() => processStore());
    render(<BottomBar />);

    fireEvent.click(screen.getByLabelText('program-button-vscode'));
    await jest.advanceTimersByTimeAsync(1000);

    expect(result.current.ActiveApp.apps[ProgramType.VSCODE]).toMatchSnapshot();
  });

  it('should render correctly github app to match snapshot', async () => {
    const { result } = renderHook(() => processStore());
    render(<BottomBar />);

    fireEvent.click(screen.getByLabelText('program-button-github'));
    await jest.advanceTimersByTimeAsync(1000);

    expect(result.current.ActiveApp.apps[ProgramType.GITHUB]).toMatchSnapshot();
  });

  it('should render correctly settings app to match snapshot', async () => {
    const { result } = renderHook(() => processStore());
    render(<BottomBar />);

    fireEvent.click(screen.getByLabelText('program-button-settings'));
    await jest.advanceTimersByTimeAsync(1000);

    expect(
      result.current.ActiveApp.apps[ProgramType.SETTINGS],
    ).toMatchSnapshot();
  });

  it('should render correctly bin app to match snapshot', async () => {
    const { result } = renderHook(() => processStore());
    render(<BottomBar />);

    fireEvent.click(screen.getByLabelText('program-button-bin'));
    await jest.advanceTimersByTimeAsync(1000);

    expect(result.current.ActiveApp.apps[ProgramType.BIN]).toMatchSnapshot();
  });
});
