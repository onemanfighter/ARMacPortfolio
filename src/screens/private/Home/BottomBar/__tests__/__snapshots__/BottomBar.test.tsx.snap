// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomBar should render correctly to match snapshot 1`] = `
<div>
  <div
    aria-label="bottom-bar"
    class="css-c60n9a"
  >
    <div
      class="css-1fa61"
      data-rbd-droppable-context-id="0"
      data-rbd-droppable-id="droppable-list"
    >
      <div
        aria-describedby="rbd-hidden-text-0-hidden-text-0"
        class="css-dvxtzn"
        data-rbd-drag-handle-context-id="0"
        data-rbd-drag-handle-draggable-id="Finder"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="Finder"
        draggable="false"
        role="button"
        tabindex="0"
      >
        <img
          alt="Finder"
          class="chakra-image css-h67rfo"
          loading="lazy"
          src=""
        />
        <div
          class="css-1es0kgi"
        />
      </div>
      <div
        aria-describedby="rbd-hidden-text-0-hidden-text-0"
        class="css-dvxtzn"
        data-rbd-drag-handle-context-id="0"
        data-rbd-drag-handle-draggable-id="Chrome"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="Chrome"
        draggable="false"
        role="button"
        tabindex="0"
      >
        <img
          alt="Chrome"
          class="chakra-image css-h67rfo"
          loading="lazy"
          src=""
        />
        <div
          class="css-1es0kgi"
        />
      </div>
      <div
        aria-describedby="rbd-hidden-text-0-hidden-text-0"
        class="css-dvxtzn"
        data-rbd-drag-handle-context-id="0"
        data-rbd-drag-handle-draggable-id="VsCode"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="VsCode"
        draggable="false"
        role="button"
        tabindex="0"
      >
        <img
          alt="VsCode"
          class="chakra-image css-h67rfo"
          loading="lazy"
          src=""
        />
        <div
          class="css-1es0kgi"
        />
      </div>
      <div
        aria-describedby="rbd-hidden-text-0-hidden-text-0"
        class="css-dvxtzn"
        data-rbd-drag-handle-context-id="0"
        data-rbd-drag-handle-draggable-id="Terminal"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="Terminal"
        draggable="false"
        role="button"
        tabindex="0"
      >
        <img
          alt="Terminal"
          class="chakra-image css-h67rfo"
          loading="lazy"
          src=""
        />
        <div
          class="css-1es0kgi"
        />
      </div>
      <div
        aria-describedby="rbd-hidden-text-0-hidden-text-0"
        class="css-dvxtzn"
        data-rbd-drag-handle-context-id="0"
        data-rbd-drag-handle-draggable-id="Spotify"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="Spotify"
        draggable="false"
        role="button"
        tabindex="0"
      >
        <img
          alt="Spotify"
          class="chakra-image css-h67rfo"
          loading="lazy"
          src=""
        />
        <div
          class="css-1es0kgi"
        />
      </div>
      <div
        aria-describedby="rbd-hidden-text-0-hidden-text-0"
        class="css-dvxtzn"
        data-rbd-drag-handle-context-id="0"
        data-rbd-drag-handle-draggable-id="Github"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="Github"
        draggable="false"
        role="button"
        tabindex="0"
      >
        <img
          alt="Github"
          class="chakra-image css-h67rfo"
          loading="lazy"
          src=""
        />
        <div
          class="css-1es0kgi"
        />
      </div>
      <div
        aria-describedby="rbd-hidden-text-0-hidden-text-0"
        class="css-dvxtzn"
        data-rbd-drag-handle-context-id="0"
        data-rbd-drag-handle-draggable-id="Settings"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="Settings"
        draggable="false"
        role="button"
        tabindex="0"
      >
        <img
          alt="Settings"
          class="chakra-image css-h67rfo"
          loading="lazy"
          src=""
        />
        <div
          class="css-1es0kgi"
        />
      </div>
    </div>
  </div>
</div>
`;
